File: android/app/src/main/kotlin/com/example/bbtotal/NativeCheckInWebViewFactory.kt
package com.example.bbtotal

import android.annotation.SuppressLint
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.view.View
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.webkit.JavascriptInterface
import android.webkit.WebChromeClient
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import org.json.JSONObject

class NativeCheckInWebViewFactory(
    private val messenger: BinaryMessenger,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(
        context: Context,
        viewId: Int,
        args: Any?,
    ): PlatformView {
        val creationParams = args as? Map<*, *> ?: emptyMap<String, Any?>()
        return NativeCheckInWebView(
            context = context,
            messenger = messenger,
            viewId = viewId,
            creationParams = creationParams,
        )
    }

    companion object {
        const val VIEW_TYPE = "bbtotal/native_check_in_webview"
    }
}

@SuppressLint("SetJavaScriptEnabled")
private class NativeCheckInWebView(
    context: Context,
    messenger: BinaryMessenger,
    viewId: Int,
    creationParams: Map<*, *>,
) : PlatformView {
    private val mainHandler = Handler(Looper.getMainLooper())
    private val methodChannel = MethodChannel(
        messenger,
        "bbtotal/native_check_in_webview/$viewId",
    )
    private val webView = WebView(context)

    private val url = creationParams.stringValue("url")
    private val allowedOriginRule = creationParams.stringValue("allowedOriginRule")
    private val locationPayloadJson =
        creationParams.stringValue("locationPayloadJson", "{}")
    private val userinfoResultJson =
        creationParams.stringValue("userinfoResultJson", "{}")
    private val wifiInfoJson =
        creationParams.stringValue("wifiInfoJson", "{}")
    private val startupScript =
        creationParams.stringValue("startupScript")

    private val supportsDocumentStartScript =
        WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)

    init {
        methodChannel.setMethodCallHandler(::onMethodCall)

        webView.layoutParams = android.widget.FrameLayout.LayoutParams(
            MATCH_PARENT,
            MATCH_PARENT,
        )
        webView.settings.javaScriptEnabled = true
        webView.settings.domStorageEnabled = true
        webView.settings.javaScriptCanOpenWindowsAutomatically = true
        webView.webChromeClient = WebChromeClient()
        webView.addJavascriptInterface(
            NativeBridgeBackend(),
            NATIVE_BRIDGE_NAME,
        )
        webView.webViewClient = object : WebViewClient() {
            override fun onPageStarted(
                view: WebView?,
                url: String?,
                favicon: android.graphics.Bitmap?,
            ) {
                sendEvent(
                    mapOf(
                        "type" to "pageStarted",
                        "url" to (url ?: ""),
                    ),
                )
                maybeInjectFallbackScript()
            }

            override fun onPageFinished(
                view: WebView?,
                url: String?,
            ) {
                maybeInjectFallbackScript()
                sendEvent(
                    mapOf(
                        "type" to "pageFinished",
                        "url" to (url ?: ""),
                    ),
                )
            }

            override fun onReceivedError(
                view: WebView?,
                request: WebResourceRequest?,
                error: WebResourceError?,
            ) {
                if (request?.isForMainFrame == false) {
                    return
                }
                sendEvent(
                    mapOf(
                        "type" to "error",
                        "description" to (
                            error?.description?.toString()
                                ?: "Unknown Android WebView error"
                            ),
                    ),
                )
            }

            @Deprecated("Deprecated in Java")
            override fun onReceivedError(
                view: WebView?,
                errorCode: Int,
                description: String?,
                failingUrl: String?,
            ) {
                sendEvent(
                    mapOf(
                        "type" to "error",
                        "description" to (description ?: "Unknown Android WebView error"),
                    ),
                )
            }
        }

        if (supportsDocumentStartScript) {
            try {
                WebViewCompat.addDocumentStartJavaScript(
                    webView,
                    startupScript,
                    setOf(allowedOriginRule),
                )
                sendLog("Android document-start injection enabled.")
            } catch (error: Exception) {
                sendLog("Android document-start injection failed: ${error.message}")
            }
        } else {
            sendLog(
                "Android WebView lacks DOCUMENT_START_SCRIPT. Falling back to early evaluateJavascript injection.",
            )
        }

        if (url.isNotBlank()) {
            webView.loadUrl(url)
        } else {
            sendEvent(
                mapOf(
                    "type" to "error",
                    "description" to "Missing WebView URL.",
                ),
            )
        }
    }

    override fun getView(): View = webView

    override fun dispose() {
        methodChannel.setMethodCallHandler(null)
        webView.stopLoading()
        webView.webChromeClient = WebChromeClient()
        webView.webViewClient = WebViewClient()
        webView.removeJavascriptInterface(NATIVE_BRIDGE_NAME)
        webView.destroy()
    }

    private fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        when (call.method) {
            "applyPreset" -> {
                applyPreset()
                result.success(null)
            }

            "reload" -> {
                mainHandler.post {
                    webView.reload()
                }
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun maybeInjectFallbackScript() {
        if (supportsDocumentStartScript || startupScript.isBlank()) {
            return
        }
        evaluateScript(startupScript)
    }

    private fun applyPreset() {
        evaluateScript(
            "window.__bbtotalApplyPreset && window.__bbtotalApplyPreset();",
        )
    }

    private fun notifyUserinfo() {
        evaluateScript(
            "window.__bbtotalNotifyUserinfo && window.__bbtotalNotifyUserinfo();",
        )
    }

    private fun notifyWifiinfo() {
        evaluateScript(
            "window.__bbtotalNotifyWifiinfo && window.__bbtotalNotifyWifiinfo();",
        )
    }

    private fun evaluateScript(script: String) {
        if (script.isBlank()) {
            return
        }
        mainHandler.post {
            webView.evaluateJavascript(script, null)
        }
    }

    private fun handleBridgeMessage(rawMessage: String?) {
        val command = decodeCommand(rawMessage) ?: run {
            if (!rawMessage.isNullOrBlank()) {
                sendLog("Ignored bridge message: $rawMessage")
            }
            return
        }

        val type = command.optString("type").ifBlank {
            command.optString("function")
        }

        when {
            type == "getLocation" || type == "getUpdatingLocation" -> applyPreset()
            type == "getUserinfo" || type == "getUserInfo" -> notifyUserinfo()
            type == "getWifiinfo" -> notifyWifiinfo()
            type == "hiddenTitle" -> sendBridgeControl("hiddenTitle")
            type == "config" -> {
                val hideNav = command.opt("hideNav")
                val shouldHide = when (hideNav) {
                    is Boolean -> hideNav
                    is String -> {
                        hideNav.equals("YES", ignoreCase = true) ||
                            hideNav.equals("true", ignoreCase = true)
                    }

                    else -> false
                }
                sendBridgeControl(if (shouldHide) "hiddenTitle" else "showTitle")
            }

            command.optBoolean("pop") -> sendBridgeControl("pop")
            command.has("openUrl") -> {
                sendEvent(
                    mapOf(
                        "type" to "bridgeControl",
                        "command" to mapOf(
                            "type" to "openUrl",
                            "url" to command.optString("openUrl"),
                        ),
                    ),
                )
            }

            type == "reloadData" -> {
                sendBridgeControl("reload")
                mainHandler.post {
                    webView.reload()
                }
            }

            type == "yuyueMeeting" ||
                type == "endMeeting" ||
                type == "joinyuyueMeeting" ||
                type == "joinyuyueZhibo" ||
                type == "launchMiniProgram" -> {
                sendLog("Native action requested: $command")
            }

            else -> sendLog("Bridge passthrough: $command")
        }
    }

    private fun decodeCommand(rawMessage: String?): JSONObject? {
        val trimmed = rawMessage?.trim().orEmpty()
        if (trimmed.isEmpty()) {
            return null
        }

        return try {
            JSONObject(trimmed)
        } catch (_: Exception) {
            null
        }
    }

    private fun sendBridgeControl(type: String) {
        sendEvent(
            mapOf(
                "type" to "bridgeControl",
                "command" to mapOf("type" to type),
            ),
        )
    }

    private fun sendLog(message: String) {
        sendEvent(
            mapOf(
                "type" to "log",
                "message" to message,
            ),
        )
    }

    private fun sendEvent(event: Map<String, Any?>) {
        mainHandler.post {
            methodChannel.invokeMethod("onEvent", event)
        }
    }

    private inner class NativeBridgeBackend {
        @JavascriptInterface
        fun getLocation(): String = locationPayloadJson

        @JavascriptInterface
        fun getUpdatingLocation(): String = locationPayloadJson

        @JavascriptInterface
        fun getUserinfo(): String = userinfoResultJson

        @JavascriptInterface
        fun getUserInfo(): String = userinfoResultJson

        @JavascriptInterface
        fun getWifiinfo(): String = wifiInfoJson

        @JavascriptInterface
        fun postMessage(message: String?) {
            handleBridgeMessage(message)
        }
    }

    companion object {
        private const val NATIVE_BRIDGE_NAME = "bbtotalNativeBridge"
    }
}

private fun Map<*, *>.stringValue(
    key: String,
    defaultValue: String = "",
): String {
    return this[key]?.toString() ?: defaultValue
}

File: ios/Runner/NativeCheckInWebViewFactory.swift
import Flutter
import UIKit
import WebKit

final class NativeCheckInWebViewFactory: NSObject, FlutterPlatformViewFactory {
  static let viewType = "bbtotal/native_check_in_webview"

  private let messenger: FlutterBinaryMessenger

  init(messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
    super.init()
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    return FlutterStandardMessageCodec.sharedInstance()
  }

  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    let creationParams = args as? [String: Any] ?? [:]
    return NativeCheckInWebView(
      frame: frame,
      messenger: messenger,
      viewId: viewId,
      creationParams: creationParams
    )
  }
}

final class NativeCheckInWebView: NSObject, FlutterPlatformView {
  private let methodChannel: FlutterMethodChannel
  private let startupScript: String
  private let urlString: String
  private let webView: WKWebView
  private let nativeBridgeHandler: WeakScriptMessageHandler
  private let syAppModelHandler: WeakScriptMessageHandler

  init(
    frame: CGRect,
    messenger: FlutterBinaryMessenger,
    viewId: Int64,
    creationParams: [String: Any]
  ) {
    self.startupScript = creationParams["startupScript"] as? String ?? ""
    self.urlString = creationParams["url"] as? String ?? ""

    let userContentController = WKUserContentController()
    userContentController.addUserScript(
      WKUserScript(
        source: startupScript,
        injectionTime: .atDocumentStart,
        forMainFrameOnly: false
      )
    )

    let configuration = WKWebViewConfiguration()
    configuration.userContentController = userContentController
    configuration.preferences.javaScriptEnabled = true

    self.webView = WKWebView(frame: frame, configuration: configuration)
    self.methodChannel = FlutterMethodChannel(
      name: "bbtotal/native_check_in_webview/\(viewId)",
      binaryMessenger: messenger
    )
    self.nativeBridgeHandler = WeakScriptMessageHandler()
    self.syAppModelHandler = WeakScriptMessageHandler()

    super.init()

    nativeBridgeHandler.delegate = self
    syAppModelHandler.delegate = self

    userContentController.add(
      nativeBridgeHandler,
      name: Self.nativeBridgeHandlerName
    )
    userContentController.add(
      syAppModelHandler,
      name: Self.syAppModelHandlerName
    )

    webView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    webView.navigationDelegate = self
    methodChannel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call: call, result: result)
    }

    if let url = URL(string: urlString) {
      webView.load(URLRequest(url: url))
    } else {
      sendEvent([
        "type": "error",
        "description": "Missing WebView URL."
      ])
    }
  }

  deinit {
    methodChannel.setMethodCallHandler(nil)
    webView.configuration.userContentController.removeScriptMessageHandler(
      forName: Self.nativeBridgeHandlerName
    )
    webView.configuration.userContentController.removeScriptMessageHandler(
      forName: Self.syAppModelHandlerName
    )
  }

  func view() -> UIView {
    return webView
  }

  private func handle(call: FlutterMethodCall, result: FlutterResult) {
    switch call.method {
    case "applyPreset":
      applyPreset()
      result(nil)
    case "reload":
      webView.reload()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func applyPreset() {
    evaluateJavaScript(
      "window.__bbtotalApplyPreset && window.__bbtotalApplyPreset();"
    )
  }

  private func notifyUserinfo() {
    evaluateJavaScript(
      "window.__bbtotalNotifyUserinfo && window.__bbtotalNotifyUserinfo();"
    )
  }

  private func notifyWifiinfo() {
    evaluateJavaScript(
      "window.__bbtotalNotifyWifiinfo && window.__bbtotalNotifyWifiinfo();"
    )
  }

  private func evaluateJavaScript(_ script: String) {
    guard !script.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
      return
    }

    DispatchQueue.main.async { [weak self] in
      self?.webView.evaluateJavaScript(script, completionHandler: nil)
    }
  }

  private func handleBridgeMessage(_ rawMessage: String?) {
    guard let command = decodeCommand(rawMessage) else {
      if let rawMessage, !rawMessage.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        sendLog("Ignored bridge message: \(rawMessage)")
      }
      return
    }

    let type = nonEmptyString(command["type"]) ?? nonEmptyString(command["function"]) ?? ""

    switch type {
    case "getLocation", "getUpdatingLocation":
      applyPreset()
    case "getUserinfo", "getUserInfo":
      notifyUserinfo()
    case "getWifiinfo":
      notifyWifiinfo()
    case "hiddenTitle":
      sendBridgeControl(type: "hiddenTitle")
    case "config":
      let hideNavValue = command["hideNav"]
      let shouldHide = boolValue(hideNavValue)
      sendBridgeControl(type: shouldHide ? "hiddenTitle" : "showTitle")
    case "reloadData":
      sendBridgeControl(type: "reload")
      webView.reload()
    case "yuyueMeeting", "endMeeting", "joinyuyueMeeting", "joinyuyueZhibo", "launchMiniProgram":
      sendLog("Native action requested: \(command)")
    default:
      if let url = nonEmptyString(command["openUrl"]) {
        sendEvent([
          "type": "bridgeControl",
          "command": [
            "type": "openUrl",
            "url": url
          ]
        ])
      } else if boolValue(command["pop"]) {
        sendBridgeControl(type: "pop")
      } else {
        sendLog("Bridge passthrough: \(command)")
      }
    }
  }

  private func decodeCommand(_ rawMessage: String?) -> [String: Any]? {
    guard
      let rawMessage,
      let data = rawMessage.data(using: .utf8),
      let object = try? JSONSerialization.jsonObject(with: data),
      let command = object as? [String: Any]
    else {
      return nil
    }

    return command
  }

  private func normalizeMessageString(body: Any) -> String? {
    if let bodyString = body as? String {
      return bodyString
    }

    guard JSONSerialization.isValidJSONObject(body) else {
      return nil
    }

    guard
      let data = try? JSONSerialization.data(withJSONObject: body),
      let bodyString = String(data: data, encoding: .utf8)
    else {
      return nil
    }

    return bodyString
  }

  private func nonEmptyString(_ value: Any?) -> String? {
    guard let string = value as? String else {
      return nil
    }
    return string.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : string
  }

  private func boolValue(_ value: Any?) -> Bool {
    switch value {
    case let boolValue as Bool:
      return boolValue
    case let stringValue as String:
      return stringValue.caseInsensitiveCompare("YES") == .orderedSame ||
        stringValue.caseInsensitiveCompare("true") == .orderedSame
    default:
      return false
    }
  }

  private func sendBridgeControl(type: String) {
    sendEvent([
      "type": "bridgeControl",
      "command": [
        "type": type
      ]
    ])
  }

  private func sendLog(_ message: String) {
    sendEvent([
      "type": "log",
      "message": message
    ])
  }

  private func sendEvent(_ event: [String: Any]) {
    DispatchQueue.main.async { [weak self] in
      self?.methodChannel.invokeMethod("onEvent", arguments: event)
    }
  }

  private static let nativeBridgeHandlerName = "bbtotalNativeBridge"
  private static let syAppModelHandlerName = "SYAppModel"
}

private final class WeakScriptMessageHandler: NSObject, WKScriptMessageHandler {
  weak var delegate: WKScriptMessageHandler?

  func userContentController(
    _ userContentController: WKUserContentController,
    didReceive message: WKScriptMessage
  ) {
    delegate?.userContentController(
      userContentController,
      didReceive: message
    )
  }
}

extension NativeCheckInWebView: WKNavigationDelegate {
  func webView(
    _ webView: WKWebView,
    didStartProvisionalNavigation navigation: WKNavigation!
  ) {
    sendEvent([
      "type": "pageStarted",
      "url": webView.url?.absoluteString ?? urlString
    ])
  }

  func webView(
    _ webView: WKWebView,
    didFinish navigation: WKNavigation!
  ) {
    sendEvent([
      "type": "pageFinished",
      "url": webView.url?.absoluteString ?? urlString
    ])
  }

  func webView(
    _ webView: WKWebView,
    didFail navigation: WKNavigation!,
    withError error: Error
  ) {
    sendEvent([
      "type": "error",
      "description": error.localizedDescription
    ])
  }

  func webView(
    _ webView: WKWebView,
    didFailProvisionalNavigation navigation: WKNavigation!,
    withError error: Error
  ) {
    sendEvent([
      "type": "error",
      "description": error.localizedDescription
    ])
  }
}

extension NativeCheckInWebView: WKScriptMessageHandler {
  func userContentController(
    _ userContentController: WKUserContentController,
    didReceive message: WKScriptMessage
  ) {
    handleBridgeMessage(normalizeMessageString(body: message.body))
  }
}
